src/fluree/db/query/fql/syntax.cljc

```diff
@@ -0,0 +1,198 @@
+
+(s/def ::orderBy (s/or :clause     ::ordering
+                       :collection (s/coll-of ::ordering)))
+(s/def ::order-by ::orderBy)
+
+(s/def ::groupBy (s/or :clause     ::var
+                       :collection (s/coll-of ::var)))
+(s/def ::group-by ::groupBy)
+
+(def first-key
+  (comp key first))
+
+(s/def ::where-op #{:filter :optional :union :bind})
+
+(defmulti where-map-spec first-key)
+
+(s/def ::where-map (s/and (s/map-of ::where-op any?, :count 1)
+                          (s/multi-spec where-map-spec first-key)))
+
+(s/def ::triple (s/cat :subject   (s/or :var variable?, :val ::subject)
+                       :predicate (s/or :var variable?, :iri ::iri)
+                       :object    (s/or :var   variable?
+                                        :ident pred-ident?
+                                        :val   any?)))
+
+(s/def ::where-tuple (s/or :triple  ::triple
+                           :binding (s/coll-of any?, :count 2)
+                           :remote  (s/coll-of any?, :count 4)))
+
+(s/def ::where-pattern (s/or :map   ::where-map
+                             :tuple ::where-tuple))
+
+(s/def ::where (s/coll-of ::where-pattern))
+
+(s/def ::optional (s/or :single     ::where-pattern
+                        :collection ::where))
+
+(s/def ::union (s/coll-of ::where))
+
+(s/def ::bind (s/map-of ::var any?))
+
+(s/def ::where (s/coll-of (s/or :map   ::where-map
+                                :tuple ::where-tuple)))
+
+(defmethod where-map-spec :filter
+  [_]
+  (s/keys :req-un [::filter]))
+
+(defmethod where-map-spec :optional
+  [_]
+  (s/keys :req-un [::optional]))
+
+(defmethod where-map-spec :union
+  [_]
+  (s/keys :req-un [::union]))
+
+(defmethod where-map-spec :bind
+  [_]
+  (s/keys :req-un [::bind]))
+
+(def never? (constantly false))
+
+(defmethod where-map-spec :minus
+  [_]
+  ;; negation - SPARQL 1.1, not yet supported
+  never?)
+
+(defmethod where-map-spec :default
+  [_]
+  never?)
+
+(s/def ::vars (s/map-of ::var any?))
+
+(s/def ::from (s/or :subj ::subject
+                    :coll (s/coll-of sid?))) ; only sids are supported for
+                                             ; specifying multiple subjects
+
+(s/def ::basic-query (s/keys :req-un [::from]))
+
+(s/def ::delete ::triple)
+
+(s/def ::delete-op (s/keys :req-un [::delete ::where]
+                           :opt-un [::vars]))
+
+(s/def ::analytical-query
+  (s/keys :req-un [::where]
+          :opt-un [::select ::selectOne ::select-one ::orderBy ::order-by ::groupBy
+                   ::group-by ::filter ::vars ::limit ::offset ::maxFuel ::max-fuel
+                   ::depth ::opts ::prettyPrint ::pretty-print]))
+
+(defn validate
+  [qry]
+  (if (s/valid? ::analytical-query qry)
+    qry
+    (throw (ex-info "Invalid Query"
+                    {:status  400
+                     :error   :db/invalid-query
+                     :reasons (s/explain-data ::analytical-query qry)}))))
```

Comment:
🤔  (Not a blocker for this PR, just an opinion) In the long-term I worry about throwing the output of `s/explain` out at people, I find them hard to parse even as a Clojure programmer (which not all users are going to be)

Reply:
Yep, I'm concerned by it too. The `explain-data` is just a temporary stop gap. I didn't want to let error messages hold up getting the query engine out, so I opened #312 to address this.